component/src/ibc/packet.rs

```diff
@@ -0,0 +1,165 @@
+use crate::ibc::component::channel::View as _;
+use crate::ibc::component::client::View as _;
+use crate::ibc::component::connection::View as _;
+use crate::Context;
+use anyhow::Result;
+use async_trait::async_trait;
+use ibc::core::ics02_client::client_state::ClientState;
+use ibc::core::ics02_client::height::Height;
+use ibc::core::ics04_channel::channel::State as ChannelState;
+use ibc::core::ics04_channel::packet::Packet;
+use ibc::core::ics24_host::identifier::ChannelId;
+use ibc::core::ics24_host::identifier::PortId;
+use penumbra_storage::StateExt;
+use penumbra_transaction::action::ICS20Withdrawal;
+
+pub trait CheckStatus: private::Sealed {}
+
+#[derive(Debug, Clone)]
+pub enum Checked {}
+impl CheckStatus for Checked {}
+
+#[derive(Debug, Clone)]
+pub enum Unchecked {}
+impl CheckStatus for Unchecked {}
+
+mod private {
+    use super::*;
+
+    pub trait Sealed {}
+    impl Sealed for Checked {}
+    impl Sealed for Unchecked {}
+}
+
+pub struct IBCPacket<S: CheckStatus> {
```

Comment:
I thought this typestate programming would be nice for correctness, but it's kind of awkward right now since there's not an easy way to persist data across a tx execution cycle (from check <> execute). So right now this reruns the check in execute, which drops some performance on the floor

Reply:
Hmm, let's make a note to come back to this once we land the state refactoring?